larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
+        bconv2d_params->padding_values.height;
+    const std::int32_t input_padding_left =
+        bconv2d_params->padding_values.width;
+    const std::int32_t batch_size = bitpacked_input_shape.Dims(0);
+    const std::int32_t input_height = bitpacked_input_shape.Dims(1);
+    const std::int32_t input_width = bitpacked_input_shape.Dims(2);
+    const std::int32_t output_height = output_shape.Dims(1);
+    const std::int32_t output_width = output_shape.Dims(2);
+    const std::int32_t output_size = num_output_pixels;
+    const std::int32_t tiled_output_size =
+        block_size_pixels *
+        ((output_size + block_size_pixels - 1) / block_size_pixels);
+
+    // Create the indirection buffer with padding (+ block_size_pixels) and fill
+    // it with pointers to the first element of the input, so that the padding
+    // at the end of the array contains pointers to valid memory.
+    indirection_buffer.assign(
+        tiled_output_size * filter_size + block_size_pixels, input_ptr);
+    // Assign the zero buffer that will be used for padding.
+    zero_buffer.assign(filter_size * input_depth, 0);
+
+    for (std::int32_t output_tile_start = 0;
+         output_tile_start < tiled_output_size;
+         output_tile_start += block_size_pixels) {
+      for (std::int32_t output_tile_offset = 0;
+           output_tile_offset < block_size_pixels; output_tile_offset++) {
+        const std::int32_t output_index =
+            std::min(output_tile_start + output_tile_offset, output_size - 1);
+        const std::int32_t batch_index =
+            output_index / (output_height * output_width);
+        const std::int32_t output_x = output_index % output_width;
+        const std::int32_t output_y =
+            (output_index % (output_height * output_width)) / output_width;
+        for (std::int32_t f_y = 0; f_y < filter_height; f_y++) {
+          const std::int32_t input_y = output_y * stride_height +
+                                       f_y * dilation_height -
+                                       input_padding_top;
+          if (0 <= input_y && input_y < input_height) {
+            for (std::int32_t f_x = 0; f_x < filter_width; f_x++) {
+              const std::int32_t input_x = output_x * stride_width +
+                                           f_x * dilation_width -
+                                           input_padding_left;
+              const std::int32_t kernel_index = f_y * filter_width + f_x;
+              const std::int32_t index = output_tile_start * filter_size +
+                                         kernel_index * block_size_pixels +
+                                         output_tile_offset;
+              if (0 <= input_x && input_x < input_width) {
+                indirection_buffer.at(index) =
+                    (input_ptr + (batch_index * input_height * input_width +
+                                  input_y * input_width + input_x) *
+                                     input_depth);
```

Comment:
Side note: we should consider increasing the clang-format line-length, that will make this function more readable :slightly_smiling_face:

Reply:
I'd be in favour of that too, but since that will touch everything let's leave that for a future PR.

Reply:
We can also do this with the `tflite::OffsetOf` helper, although I'm not sure how much it will improve readability. It will be something like `OffsetOf(input_shape, batch_index, input_y, input_x, 0);`